File: app/javascript/channels/room_channel.js
import consumer from "./consumer"

const room = (setRoomNames) => consumer.subscriptions.create("RoomChannel", {
  connected() {
    // Called when the subscription is ready for use on the server
    console.log("Connected to the room channel");
  },

  disconnected() {
    // Called when the subscription has been terminated by the server
  },

  received(data) {
    // Called when there's incoming data on the websocket for this channel
    console.log("Received data: ", data);
    if (data) setRoomNames(data);
  },

  send_message: function(name) {
    return this.perform('send_message', { message: name });
  }
});

export default room;

File: app/channels/room_channel.rb
class RoomChannel < ApplicationCable::Channel
  def subscribed
    # stream_from "some_channel"
    stream_from "room_channel"
    ActionCable.server.broadcast("room_channel", ChatChannel::CHATROOMS.keys)
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def send_message(data)
    keys = ChatChannel::CHATROOMS.keys
    if !keys.include?(data['message'])
      keys << data['message']
      ChatChannel::CHATROOMS[data['message']] ||= []
    end
    ActionCable.server.broadcast("room_channel", keys)
  end
end
